educe/stac/oneoff/weave.py

```diff
@@ -0,0 +1,267 @@
+    source document.
+    """
+    gaps = []
+    last_idx = 0
+    for src, _, size in matches:
+        if src != last_idx:
+            gaps.append((last_idx, src - last_idx))
+        last_idx = src + size
+    return gaps
+
+
+def tgt_gaps(matches):
+    """
+    Given matches between the source and target document, return the spaces
+    between these matches as target offset and size (a bit like the matches).
+    By rights this should be empty, but you never know
+    """
+    gaps = []
+    last_idx = 0
+    for _, tgt, size in matches:
+        if tgt != last_idx:
+            gaps.append((last_idx, tgt - last_idx))
+        last_idx = tgt + size
+    return gaps
+
+
+def check_matches(tgt_doc, matches):
+    """
+    Check that the target document text is indeed a subsequence of
+    the source document text (the source document is expected to be
+    "augmented" version of the target with new text interspersed
+    throughout)
+    """
+    tgt_text = tgt_doc.text()
+
+    if not tgt_text:
+        return
+    elif not matches:
+        raise WeaveException('no matches in non-empty target doc')
+    elif matches[0].b != 0:
+        oops = ('matches ({}) do not start at beginning of target '
+                'document <{}>').format(matches[0], tgt_doc.origin)
+        raise WeaveException(oops)
+
+    gaps = tgt_gaps(matches)
+    if gaps:
+        oops = 'there are match gaps in the target document: {}'
+        raise WeaveException(oops.format(gaps))
+
+    _, tgt, size = matches[-1]
+    if tgt + size != len(tgt_text):
+        raise WeaveException('matches do not cover the full target '
+                             'document')
+
+
+def _units_between(doc, start, end):
+    """Unit-level annotations in a span defined by a starting point
+    and its width
+
+    Parameters
+    ----------
+    doc : Document
+    start : int
+    end : end
+
+    Returns
+    -------
+    annos
+        list of annotation objects
+    """
+    return enclosed(Span(start, end), doc.units)
+
+
+def compute_updates(src_doc, tgt_doc, matches):
+    """Return updates that would need to be made on the target
+    document.
+
+    Given matches between the source and target document, return span
+    updates along with any source annotations that do not have an
+    equivalent in the target document (the latter may indicate that
+    resegmentation has taken place, or that there is some kind of problem)
+
+    Parameters
+    ----------
+    src_doc : Document
+    tgt_doc : Document
+    matches : [Match]
+
+    Returns
+    -------
+    updates: Updates
+    """
+    res = Updates()
+
+    # case 2 and 5 (to be pruned below)
+    res.expected_src_only.extend(src_doc.units)
+    res.abnormal_tgt_only.extend(tgt_doc.units)
+
+    # case 1, 2 and 4
+    for src, tgt, size in matches:
+        tgt_to_src = src - tgt
+        res.shift_if_ge[tgt] = tgt_to_src  # case 1 and 2
+        src_annos = _units_between(src_doc, src, src + size)
```

Comment:
Could you unroll `_units_between` in these two lines and kill the said function?
It is a simple one-liner that calls another simple one-liner (`educe.stac.context.enclosed`) that itself calls what could be (written as) a simple one-liner (`educe.annotation.Span.encloses`). This is usually a bad sign.